lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_instance.py

```diff
@@ -0,0 +1,381 @@
+                msg="failed to delete dynamic resource instance %s: %s" % (
+                    dynamic_resource_instance.get('name'), str(e)
+                )
+            )
+
+        return dict(
+            changed=True,
+            msg="successfully deleted the dynamic resource instance %s" % (
+                dynamic_resource_instance.get('name')
+            )
+        )
+
+
+def main():
+    argument_spec = dict(
+        name=dict(required=True, type='str'),
+        state=dict(choices=['absent', 'present'], default='present'),
+        dynamic_resource_definition=dict(type='str'),
+        property_attributes=dict(type='dict'),
+        services=dict(type='list', default=[]),
+        providers=dict(type='list', default=[]),
+    )
+
+    required_if = [
+        (
+            'state', 'present', (
+                'dynamic_resource_definition', 'property_attributes',
+            )
+        )
+    ]
+    # add the manageiq connection arguments to the arguments
+    argument_spec.update(manageiq_argument_spec())
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_if=required_if,
+    )
+
+    name = module.params['name']
+    state = module.params['state']
+    dynamic_resource_definition = module.params['dynamic_resource_definition']
+    property_attributes = module.params['property_attributes']
+    services = module.params['services']
+    providers = module.params['providers']
+
+    manageiq = ManageIQ(module)
+    manageiq_dynamic_resource_instance = ManageIQDynamicResourceInstance(manageiq)
+
+    # Check if definition exists
+    dynamic_resource_definition = manageiq_dynamic_resource_instance.dynamic_resource_definition(
+        dynamic_resource_definition
+    )
+    if not dynamic_resource_definition and state != "absent":
+        module.fail_json(
+            msg="Dymamic resource definition:%s does not exist." % (dynamic_resource_definition)
+        )
+    # Get instance if exists
+    dynamic_resource_instance = manageiq_dynamic_resource_instance.dynamic_resource_instance(
+        name
+    )
+
+    if state == "absent":
+        # if we have a instance, delete it
+        if dynamic_resource_instance:
+            res_args = manageiq_dynamic_resource_instance.delete_instance(
+                dynamic_resource_instance
+            )
+        # if we do not have a instance, nothing to do
+        else:
+            res_args = dict(
+                changed=False,
+                msg="dynamic resource instance  %s: does not exist in manageiq" % (name)
+            )
+
+    if state == "present":
+        # Validate that providers and services exist
+        service_objs = [
+            manageiq_dynamic_resource_instance.service(service) for service in services
+        ]
+        provider_objs = [
+            manageiq_dynamic_resource_instance.provider(provider) for provider in providers
+        ]
+
+        if not all(service_objs) and any(services):
```

Comment:
I think this can be written simpler (to understand) `if None in service_objs`.

But also the error message could be more actionable if it told you *which* ones didn't exist.
Hmm, a bit tedious:
```
service_objs = []
missing_services = []
for service_name in services:
    service_obj = manageiq_dynamic_resource_instance.service(service)  # or inlined per above comment
    if service_obj:
        service_objs.append(service_obj)
    else:
        missing_services.append(service_name)
```
you need it twice, and could be generally useful, but not sure how to generalize to module_utils...

Reply:
I think I can just add a comment to explain what this line does.
EX: Check if any of the services do not exist in manageiq